src/cls/rbd/cls_rbd.cc

```diff
@@ -2807,7 +2807,29 @@ int snapshot_set_limit(cls_method_context_t hctx, bufferlist *in,
     return -EINVAL;
   }
 
-  if (new_limit == UINT64_MAX) {
+  int max_read = RBD_MAX_KEYS_READ;
+  size_t snap_count = 0;
+  string last_read = RBD_SNAP_KEY_PREFIX;
+  bool more;
+
+  do {
+    set<string> keys;
+    rc = cls_cxx_map_get_keys(hctx, last_read, max_read, &keys, &more);
+    if (rc < 0)
+      return rc;
+    for (auto& key : keys) {
+      if (key.find(RBD_SNAP_KEY_PREFIX) != 0)
+        break;
+      snap_count++;
+    }
+    if (!keys.empty())
+      last_read = *(keys.rbegin());
+  } while (more);
+
+  if (new_limit < snap_count){
+    rc = -ERANGE;
+    CLS_ERR("error setting snapshot limit: %s", cpp_strerror(rc).c_str());
```

Comment:
The error log should really be on line 2819 (text adjusted to "error retrieving snapshots") -- this should just be a debug log message saying "snapshot limit is less than the number of snapshots".